### DISCUSSION TITLE
Process request data in other thread

### DISCUSSION THREAD

**Discussion by Neo-Ciber94** [rwf2/Rocket#2539] [Questions]
Most of the tokio types require a non-static lifetime but I'm having some problems with that when trying to move data to other thread or using tokio.

In this minimal example I want to read the request data as a stream and process it in other thread, is there a way to do it? Its not currently possible because cannot be borrowed for static.

```rust
#[rocket::get("/")]
fn handler(data: rocket::Data<'_>) {
    let data_stream = data.open(ByteUnit::max_value());
    rocket::tokio::spawn(async move {
        let reader = ReaderStream::new(data_stream);
        // to things in other in other thread
    });
}
```

**Comment by SergioBenitez** [comment]
The approach you're presenting appears, at first glance, to be conceptually mistaken. Perhaps the best way for me to respond it to present a few points for your consideration. Hopefully they collectively resolve your question.

* Request data is _always_ streamed, whether you do it directly or whether you go through existing data guards like `String` or `Form`.
* Request data must be completely read before a response can be sent. It is not possible, considering HTTP in general (since this _is_ somewhat possible in HTTP2), to send a response without reading the full request data. So reading the request in a different thread, even if accomplishable, would not improve performance from the client's perspective.
* `tokio::spawn` doesn't start a new thread but rather an async task. It runs the task in the background, and the task may outlive the parent that spawned it. As such, it cannot possible hold references to the parent or we would risk dangling pointers (the parent being dropped before the child task). Since `Data` holds a reference to the request, you cannot move it into the new task. This is by design: Rocket is trying to prevent you from doing what you're doing.
* Rocket serves all requests in separate async tasks, concurrently and in parallel. So, reading the request data asynchronously, directly in a handler, will not block the server in any way, and is the most correct and performant approach.

**Reply by Neo-Ciber94** [reply]
Thanks for the response,

What I'm trying to do is using Rocket as a type of proxy,

> Request data must be completely read before a response can be sent. It is not possible, considering HTTP in general (since this is somewhat possible in HTTP2), to send a response without reading the full request data. So reading the request in a different thread, even if accomplishable, would not improve performance from the client's perspective.

but I didn't knew the response need to be completely read in most cases, for example when uploading a large file, because of that I was avoiding buffering the body into a `Vec<u8>` or something similar, and was trying to forward the body as a stream.
